Description: 适配不同内核版本 (Velinux/OCOS/OpenEuler)
 1. feat: adapt to different kernels (c72c61c)
 2. Bug: 修复KVM页表遍历中level异常判断的类型转换缺陷 (b102206)
Author: Debian Packager
Last-Update: 2026-07-29

---

diff --git a/plugins/smap/smap.spec b/plugins/smap/smap.spec
index f2cab1a..c8cba77 100644
--- a/plugins/smap/smap.spec
+++ b/plugins/smap/smap.spec
@@ -29,9 +29,10 @@ This package contains the Huawei SMAP Driver
 %setup -q -T -b 0 -c -n smap
 
 %build
-cd %{_builddir}/smap/src/drivers && make -j`nproc`
+%{!?KERNEL_VERSION:%global KERNEL_VERSION openeuler}
+cd %{_builddir}/smap/src/drivers && make KERNEL_VERSION=%{KERNEL_VERSION} -j`nproc`
 cp %{_builddir}/smap/src/drivers/Module.symvers %{_builddir}/smap/src/tiering/depends
-cd %{_builddir}/smap/src/tiering && make -j`nproc`
+cd %{_builddir}/smap/src/tiering && make KERNEL_VERSION=%{KERNEL_VERSION} -j`nproc`
 cd %{_builddir}/smap/src/ucache && make -j`nproc`
 cd %{_builddir}/smap && cmake -DCMAKE_BUILD_TYPE=Release %{_builddir}/smap
 make -j`nproc` install
diff --git a/plugins/smap/src/drivers/Makefile b/plugins/smap/src/drivers/Makefile
index fca9738..6fa487b 100644
--- a/plugins/smap/src/drivers/Makefile
+++ b/plugins/smap/src/drivers/Makefile
@@ -12,6 +12,14 @@ smap_histogram_tracking-objs := ub_hist.o
 
 ccflags-y += -DACPI_NOT_READY
 
+ifeq ($(KERNEL_VERSION), ocos)
+    ccflags-y += -DKERNEL_OCOS
+else ifeq ($(KERNEL_VERSION), velinux)
+    ccflags-y += -DKERNEL_VELINUX
+else
+    ccflags-y += -DKERNEL_OPENEULER
+endif
+
 .PHONY: modules clean
 .default: modules
 
diff --git a/plugins/smap/src/drivers/access_tracking_wrapper.c b/plugins/smap/src/drivers/access_tracking_wrapper.c
index 67bdeec..59677cd 100644
--- a/plugins/smap/src/drivers/access_tracking_wrapper.c
+++ b/plugins/smap/src/drivers/access_tracking_wrapper.c
@@ -161,13 +161,21 @@ pte_t smap_huge_ptep_get(pte_t *ptep)
 {
 	int ncontig, i;
 	size_t pgsize;
+#ifdef KERNEL_VELINUX
+	pte_t orig_pte = ptep_get(ptep);
+#else
 	pte_t orig_pte = __ptep_get(ptep);
+#endif
 	if (!pte_present(orig_pte) || !pte_cont(orig_pte))
 		return orig_pte;
 
 	ncontig = num_contig_ptes(page_size(pte_page(orig_pte)), &pgsize);
 	for (i = 0; i < ncontig; i++, ptep++) {
+#ifdef KERNEL_VELINUX
+		pte_t pte = ptep_get(ptep);
+#else
 		pte_t pte = __ptep_get(ptep);
+#endif
 		if (pte_dirty(pte))
 			orig_pte = pte_mkdirty(orig_pte);
 		if (pte_young(pte))
diff --git a/plugins/smap/src/drivers/accessed_bit.c b/plugins/smap/src/drivers/accessed_bit.c
index 5e9492e..b92ef2f 100644
--- a/plugins/smap/src/drivers/accessed_bit.c
+++ b/plugins/smap/src/drivers/accessed_bit.c
@@ -1337,8 +1337,13 @@ static int check_pte_young(pte_t *pte, unsigned long addr, unsigned long next,
 		if (pte_walk->type == STATISTIC_SCAN)
 			pte_walk->statistic_vaddr[pte_walk->statistic_cnt++] =
 				addr;
-		if (!is_file_or_shared_page(page))
+		if (!is_file_or_shared_page(page)) {
+#ifdef KERNEL_VELINUX
+			__ptep_test_and_clear_young(pte);
+#else
 			__ptep_test_and_clear_young(NULL, 0, pte);
+#endif
+		}
 		pte_walk->flag = true;
 	}
 
@@ -1604,7 +1609,11 @@ static int pte_pure_clear(pte_t *pte, unsigned long addr, unsigned long next,
 
 	if (pte_young(*pte)) {
 		pte_walk->flag = true;
+#ifdef KERNEL_VELINUX
+		__ptep_test_and_clear_young(pte);
+#else
 		__ptep_test_and_clear_young(NULL, 0, pte);
+#endif
 	}
 
 	return 0;
diff --git a/plugins/smap/src/tiering/Makefile b/plugins/smap/src/tiering/Makefile
index 9d099e5..cdc0352 100644
--- a/plugins/smap/src/tiering/Makefile
+++ b/plugins/smap/src/tiering/Makefile
@@ -1,5 +1,9 @@
 MODULE_NAME = smap_tiering
-$(MODULE_NAME)-objs := rmap.o ksm.o smap_migrate_wrapper.o smap_migrate_pages.o migrate_task.o pid_ioctl.o migrate_back_debugfs.o work.o numa.o acpi_helper.o acpi_mem.o mig_init.o tracking_manage.o smap_debugfs.o  dump_info.o iomem.o ham_migration.o basic.o coherence_maintain.o ham_tasks_mgr.o
+$(MODULE_NAME)-objs := rmap.o ksm.o smap_migrate_wrapper.o smap_migrate_pages.o migrate_task.o pid_ioctl.o migrate_back_debugfs.o work.o numa.o acpi_helper.o acpi_mem.o mig_init.o tracking_manage.o smap_debugfs.o  dump_info.o iomem.o
+
+ifeq ($(filter $(KERNEL_VERSION), ocos velinux),)
+$(MODULE_NAME)-objs += ham_migration.o basic.o coherence_maintain.o ham_tasks_mgr.o
+endif
 
 obj-m   := $(MODULE_NAME).o
 
@@ -10,6 +14,14 @@ KBUILD_EXTRA_SYMBOLS = $(src)/depends/Module.symvers
 ifeq ($(CRITICAL), 0)
 	ccflags-y := -DCRITICAL_OFF
 endif
+ifeq ($(KERNEL_VERSION), ocos)
+    ccflags-y += -DKERNEL_OCOS
+else ifeq ($(KERNEL_VERSION), velinux)
+    ccflags-y += -DKERNEL_VELINUX
+else
+    ccflags-y += -DHAM_ENABLED
+    ccflags-y += -DKERNEL_OPENEULER
+endif
 
 .PHONY: modules
 default: modules
@@ -18,4 +30,4 @@ modules:
 	$(MAKE) -C $(KERNELDIR) M=$(PWD) modules
 
 clean:
-	$(MAKE) -C $(KERNELDIR) M=$(PWD) clean
\ No newline at end of file
+	$(MAKE) -C $(KERNELDIR) M=$(PWD) clean
diff --git a/plugins/smap/src/tiering/basic.c b/plugins/smap/src/tiering/basic.c
index 9f46517..113e6e5 100644
--- a/plugins/smap/src/tiering/basic.c
+++ b/plugins/smap/src/tiering/basic.c
@@ -43,14 +43,22 @@ pte_t kernel_huge_ptep_get(pte_t *ptep)
 {
 	int ncontig, i;
 	size_t pgsize;
+#ifdef KERNEL_VELINUX
+	pte_t orig_pte = ptep_get(ptep);
+#else
 	pte_t orig_pte = __ptep_get(ptep);
+#endif
 	if (!pte_present(orig_pte) || !pte_cont(orig_pte))
 		return orig_pte;
 
 	ncontig =
 		kernel_num_contig_ptes(page_size(pte_page(orig_pte)), &pgsize);
 	for (i = 0; i < ncontig; i++, ptep++) {
+#ifdef KERNEL_VELINUX
+		pte_t pte = ptep_get(ptep);
+#else
 		pte_t pte = __ptep_get(ptep);
+#endif
 		if (pte_dirty(pte))
 			orig_pte = pte_mkdirty(orig_pte);
 
diff --git a/plugins/smap/src/tiering/mig_init.c b/plugins/smap/src/tiering/mig_init.c
index 856b5d4..5b3ebd9 100644
--- a/plugins/smap/src/tiering/mig_init.c
+++ b/plugins/smap/src/tiering/mig_init.c
@@ -19,7 +19,6 @@
 #include "common.h"
 #include "acpi_mem.h"
 #include "iomem.h"
-#include "ham_migration.h"
 #include "smap_migrate_pages.h"
 #ifdef CRITICAL_OFF
 #include "critical.h"
diff --git a/plugins/smap/src/tiering/smap_migrate_pages.c b/plugins/smap/src/tiering/smap_migrate_pages.c
index 55d2394..4630851 100644
--- a/plugins/smap/src/tiering/smap_migrate_pages.c
+++ b/plugins/smap/src/tiering/smap_migrate_pages.c
@@ -71,6 +71,7 @@ static struct migrate_node {
 	unsigned long nr[SMAP_MAX_NUMNODES];
 } migrate_node;
 
+#ifdef CONFIG_MIGRATE_PAGES_DMA_OFFLOADING
 void set_remote_migrate_mode(unsigned int mode)
 {
 	if (mode) {
@@ -80,6 +81,11 @@ void set_remote_migrate_mode(unsigned int mode)
 	}
 	pr_info("set remote migrate mode: %u\n", remote_migrate_mode);
 }
+#else
+void set_remote_migrate_mode(unsigned int mode)
+{
+}
+#endif
 
 static int smap_check_huge_page_for_migration(struct page *page, pid_t pid)
 {
diff --git a/plugins/smap/src/tiering/tracking_manage.c b/plugins/smap/src/tiering/tracking_manage.c
index 4236835..130e7e4 100644
--- a/plugins/smap/src/tiering/tracking_manage.c
+++ b/plugins/smap/src/tiering/tracking_manage.c
@@ -26,7 +26,9 @@
 #include "dump_info.h"
 #include "acpi_mem.h"
 #include "iomem.h"
+#ifdef HAM_ENABLED
 #include "ham_migration.h"
+#endif
 #include "pid_ioctl.h"
 #ifdef CRITICAL_OFF
 #include "critical.h"
@@ -91,7 +93,9 @@ static void resource(void)
 	exit_migrate();
 	smap_dev_exit();
 	smap_debugfs_mod_exit();
+#ifdef HAM_ENABLED
 	ham_exit();
+#endif
 }
 
 static int is_qemu_name_valid(void)
@@ -238,17 +242,21 @@ static int __init tracking_init(void)
 		pr_err("failed to init migrate, ret: %d\n", ret);
 		goto out_dev_int;
 	}
+#ifdef HAM_ENABLED
 	ret = ham_init();
 	if (ret < 0) {
 		pr_err("failed to init HAM, ret: %d\n", ret);
 		goto out_migrate_int;
 	}
+#endif
 	queue_delayed_work(migrate_back_wq, &migrate_back_work,
 			   msecs_to_jiffies(MB_INTV));
 	pr_info("SMAP init successfully\n");
 	return 0;
+#ifdef HAM_ENABLED
 out_migrate_int:
 	exit_migrate();
+#endif
 out_dev_int:
 	smap_dev_exit();
 out_debugfs:
diff --git a/plugins/smap/test/CMakeLists.txt b/plugins/smap/test/CMakeLists.txt
index 6fed2d5..fbf0afa 100644
--- a/plugins/smap/test/CMakeLists.txt
+++ b/plugins/smap/test/CMakeLists.txt
@@ -15,7 +15,7 @@ set(CMAKE_SHARED_LINKER_FLAGS " -lgcov --coverage ")
 
 add_subdirectory(3rdparty)
 
-add_compile_definitions(USE_DT)
+add_compile_definitions(USE_DT CONFIG_MIGRATE_PAGES_DMA_OFFLOADING)
 
 # include all source files
 aux_source_directory(${CMAKE_CURRENT_LIST_DIR}/tiering TEST_SRC)
diff --git a/plugins/smap/test/user/manage/test_manage.cpp b/plugins/smap/test/user/manage/test_manage.cpp
index d307a88..6592f57 100644
--- a/plugins/smap/test/user/manage/test_manage.cpp
+++ b/plugins/smap/test/user/manage/test_manage.cpp
@@ -1005,8 +1005,11 @@ TEST_F(ManageTest, TestGetNumaNodesForPid)
 {
     int pid = 1234;
     int node = 0;
+    MOCKER(sched_getaffinity).stubs().will(returnValue(-1));
     int ret = GetNumaNodesForPid(pid, &node);
     EXPECT_EQ(ret, -EINVAL);
+
+    GlobalMockObject::verify();
     MOCKER(sched_getaffinity).stubs().will(returnValue(0));
     MOCKER(GetNodeFromCpu).stubs().will(returnValue(-EINVAL)).then(returnValue(0));
     ret = GetNumaNodesForPid(pid, &node);

diff --git a/plugins/smap/src/drivers/kvm_pgtable.c b/plugins/smap/src/drivers/kvm_pgtable.c
index df3223d..ca71c3d 100644
--- a/plugins/smap/src/drivers/kvm_pgtable.c
+++ b/plugins/smap/src/drivers/kvm_pgtable.c
@@ -49,7 +49,11 @@ static u32 kvm_pgtable_idx(struct kvm_pgtable_walk_data *data, u32 level)
 
 static bool kvm_pte_table(kvm_pte_t pte, u32 level)
 {
+#ifdef KERNEL_VELINUX
+	if (level == KVM_PGTABLE_LAST_LEVEL || !kvm_pte_valid(pte))
+#else
 	if (level == KVM_PGTABLE_MAX_LEVELS - 1 || !kvm_pte_valid(pte))
+#endif
 		return false;
 
 	return FIELD_GET(KVM_PTE_TYPE, pte) == KVM_PTE_TYPE_TABLE;
@@ -167,7 +171,11 @@ static int __kvm_pgtable_walk(struct kvm_pgtable_walk_data *data,
 	u32 idx;
 	int ret = 0;
 
+#ifdef KERNEL_VELINUX
+	if (WARN_ON_ONCE(level > KVM_PGTABLE_LAST_LEVEL))
+#else
 	if (WARN_ON_ONCE(level >= KVM_PGTABLE_MAX_LEVELS))
+#endif
 		return -EINVAL;
 
 	for (idx = kvm_pgtable_idx(data, level); idx < PTRS_PER_PTE; ++idx) {
